MiniApp/Classes/Storage/MiniAppKeyChain.swift

```diff
@@ -0,0 +1,124 @@
+import Foundation
+
+@objc class MiniAppKeyChain: NSObject {
+    let service: String
+    var account: String
+
+    typealias KeysDictionary = [String: [MASDKCustomPermissionModel]]
+
+    init(service: String = Bundle.main.bundleIdentifier!) {
+        self.service = service
+        self.account = "\(service).rakuten.tech.permission.keys"
+    }
+
+    func getCustomPermissions(forMiniApp keyId: String) -> [MASDKCustomPermissionModel]? {
+        guard let allKeys = keys() else {
+            return nil
+        }
+        guard let permissionList = allKeys[keyId] as [MASDKCustomPermissionModel]? else {
+            return setDefaultPermissionsInKeyChain(forMiniApp: keyId)
+        }
+        return permissionList
+    }
+
+    func storeCustomPermissions(permissions: [MASDKCustomPermissionModel], forMiniApp keyId: String) {
+        var keysDic = keys()
+        guard var cachedPermissions = self.getCustomPermissions(forMiniApp: keyId) else {
+            return
+        }
+
+        _ = permissions.map { (permissionModel: MASDKCustomPermissionModel) -> MASDKCustomPermissionModel in
+            if let index = cachedPermissions.firstIndex(of: permissionModel) {
+                cachedPermissions[index] = permissionModel
+                cachedPermissions[index].permissionDescription = ""
+            }
+            return permissionModel
+        }
+
+        if keysDic != nil {
+            keysDic?[keyId] = cachedPermissions
+        } else {
+            keysDic = [keyId: cachedPermissions]
+        }
+
+        if let keys = keysDic {
+            write(keys: keys)
+        }
+    }
+
+    internal func setDefaultPermissionsInKeyChain(forMiniApp id: String) -> [MASDKCustomPermissionModel] {
+        var supportedPermissionList = [MASDKCustomPermissionModel]()
+        MiniAppCustomPermissionType.allCases.forEach {
+            supportedPermissionList.append(MASDKCustomPermissionModel(
+                permissionName: MiniAppCustomPermissionType(
+                    rawValue: $0.rawValue)!,
+                isPermissionGranted: .denied,
+                permissionRequestDescription: ""
+            ))
+        }
+        write(keys: [id: supportedPermissionList])
+        return supportedPermissionList
+    }
+
+    private func write(keys: KeysDictionary) {
```

Comment:
Is there any way you could use the [KeyStore from iOS SDK Utils
](https://github.com/rakutentech/ios-sdkutils/blob/master/RSDKUtils/KeyStore/KeyStore.swift) to simplify this class? It seems like the writing and reading functionality is similar. We're already using the `KeyStore` class for [cache verification](https://github.com/rakutentech/ios-miniapp/blob/master/MiniApp/Classes/Storage/MiniAppCacheVerifier.swift).

Sorry, I'm not very familiar with iOS KeyChain, so I'm not sure if this makes sense to use it or not.

Reply:
That's right. I have used the most of the Keystore wrapper from SDK Utils, but our SDK needs a different dictionary [String: [MASDKCustomPermissionModel]] to be stored, so I thought I could write a similar interface in our SDK.

Reply:
Ah, okay it's slightly different. And it looks like the `KeyStore` class doesn't currently let you retrieve keys. Maybe in the future we can have a technical improvement to make `KeyStore` more generic so that it can be used in this class.

Reply:
That's the idea. I was even thinking to improve this class (Adding more interface)to take care of the Hashing technique as well.